cmd/ro/roagent/roagent.go

```diff
@@ -0,0 +1,198 @@
+// Package roagent provides ROAgent, which implements the SSH agent interface,
+// forwarding sign requests to a Red October server
+
+package roagent
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/subtle"
+	"encoding/json"
+	"errors"
+	"log"
+	"io"
+
+	roclient "github.com/cloudflare/redoctober/client"
+	"github.com/cloudflare/redoctober/core"
+	"golang.org/x/crypto/ssh"
+)
+
+type ROAgent struct {
+	locked   bool
+	keyring  []*ROSigner
+
+	server   *roclient.RemoteServer
+	username string
+	password string
+}
+
+type ROSigner struct {
+	pub     ssh.PublicKey
+	encKey  []byte
+	roagent *ROAgent
+	comment string
+}
+
+func (rosigner ROSigner) PublicKey() ssh.PublicKey {
+	return rosigner.pub
+}
+
+func (rosigner ROSigner) Sign(rand io.Reader, msg []byte) (signature *ssh.Signature, err error) {
+	req := core.SSHSignWithRequest{
+		Name:     rosigner.roagent.username,
+		Password: rosigner.roagent.password,
+		Data:     rosigner.encKey,
+		TBSData:  msg,
+	}
+
+	resp, err := rosigner.roagent.server.SSHSignWith(req)
+	if err != nil {
+		return nil, err
+	}
+	if resp.Status != "ok" {
+		return nil, errors.New("ro-ssh-agent: response status error: " + resp.Status)
+	}
+
+	var respMsg core.SSHSignatureWithDelegates
+	err = json.Unmarshal(resp.Response, &respMsg)
+	if err != nil {
+		return nil, err
+	}
+	sshSignature := ssh.Signature{
+		Format: respMsg.SignatureFormat,
+		Blob:   respMsg.Signature,
+	}
+	return &sshSignature, nil
+}
+
+// NewROAgent creates a new SSH agent which forwards signature requests to the
+// provided remote server
+func NewROAgent(server *roclient.RemoteServer, username, password string) *ROAgent {
+	return &ROAgent{
+		server:   server,
+		username: username,
+		password: password,
+		keyring:  []*ROSigner{},
+	}
+}
+
+// NewROSigner adds a new SSH identity to the ROAgent
+func NewROSigner(pubKey ssh.PublicKey, encBytes []byte) *ROSigner {
+	return &ROSigner{
+		pub:    pubKey,
+		encKey: encBytes,
+	}
+}
+
+// RemoveAll empties ROAgent's keyring
+func (roagent *ROAgent) RemoveAll() error {
+	if roagent.locked {
+		return errLocked
+	}
+	roagent.keyring = []*ROSigner{}
+	return nil
+}
+
+// Removes the first matching key from ROAgent's keyring
+func (roagent *ROAgent) Remove(key ssh.PublicKey) error {
+	if roagent.locked {
+		return errLocked
+	}
+	wanted := key.Marshal()
+	for i, rosigner := range roagent.keyring {
+		if bytes.Equal(rosigner.PublicKey().Marshal(), wanted) {
+			// Order is not preserved
+			roagent.keyring[i] = roagent.keyring[0] 
+			roagent.keyring = roagent.keyring[1:]
+			log.Println("ro-ssh-agent: signer removed")
+			return nil
+		}
+	}
+	return errors.New("ro-ssh-agent: could not remove signer")
+}
+
+// Locks the ROAgent by removing the password
+// TODO should this encrypt the password instead?
```

Comment:
Encrypt with which key?

Reply:
Removing the password effectively disables the keyring since RO would stop responding, but it might make more sense to use the entered password to symmetrically encrypt the RO password and keyring information. I'm not sure what is the behavior of ssh-agent when locking the keyring either.